Some small fixes

tanaes opened the pull request:
I came across a few errors when running some real datasets on Barnacle, which this PR fixes. 

1. There was an intermittent memory allocation exceeded error caused by the host filtering step. I tracked this down to the piped commands giving bam output to the sort, which I think forced it to load the entire output into memory prior to sorting. Instead outputting to localscratch on disk prior to sort and fastq conversion. 

2. This error would cause a befuddling Snakemake failure where the job would fail, but Snakemake wouldn't notice it. As jobs failed, this eventually caused the Snakemake process to stop submitting any new ones because it thought the failed jobs were still running. To fix, I took advantage of a new functionality in Snakemake ≥ 4.1 which allows a job status script to check in on jobs directly from the scheduler. This way Snakemake can monitor the job without writing temp files + restart as necessary. 

Note that this requires a newer version of Snakemake than we were running before, so please reinstall your base oecophylla environment.

tanaes commented:
Also note that this follows PR #119 so please merge first!

qiyunzhu commented on cluster_configs/barnacle/barnacle_status.py:
```diff
@@ -0,0 +1,61 @@
+
+    return(status)
+
+def print_status(status):
```
I kind-of feel that a function should better return a value rather than interact with the user. Instead, `main()` could print what is return by this function.

tanaes replied:
That makes sense...

qiyunzhu commented on cluster_configs/barnacle/barnacle_status.py:
```diff
@@ -0,0 +1,61 @@
+    for line in lines:
+        if line.startswith(' '):
+            entry = line.rstrip()
+            p = re.compile('^ +(.+?) = (.+)')
```
I think this line should go above the for loop, so as to increase performance. A compiled regex can be reused over and over.

tanaes replied:
Cool, did not know that!

sjanssen2 replied:
did that change happen?

tanaes replied:
Ooops missed that one.

qiyunzhu commented on cluster_configs/barnacle/barnacle_status.py:
```diff
@@ -0,0 +1,61 @@
+    except KeyError:
+        print("failed")
+
+def main():
```
Looks like there is lots of non-standard things here: `sys.argv` instead of `click`, absence of docstring, hard-coded constants, absence of unit test. I am totally fine with it though. Just a reminder.

tanaes replied:
this is a helper script for Snakemake and so doesn't interact with user---it has very prescribed inputs (one positional argument) and outputs (one of 'success', 'running', or 'failure'), so I'd rather rely on on standard libraries than Click. It is specific to Barnacle, but mostly follows [this example](https://github.com/Snakemake-Profiles/surfsara-grid/blob/master/%7B%7Bcookiecutter.profile_name%7D%7D/grid-status.py) from Johannes Köster. 

But the unit test thing is true, I do need to do that.

qiyunzhu replied:
Makes sense!

tanaes commented:
@qiyunzhu addressed your comments. This new script status *really* helps with reliability of execution on the cluster, working on porting over to SLURM too for Comet.

@mortonjt or @sjanssen2 want to take a quick look as well?

sjanssen2 commented on cluster_configs/barnacle/barnacle_status.py:
```diff
@@ -0,0 +1,65 @@
+    TRY_TIMES = 3
+
+    for i in range(TRY_TIMES):
+        cmd = "qstat -f {}".format(jobid)
```
where is format() defined? Is that an alternative way of saying "qstat -f %s" % jobid ?

tanaes replied:
yep, it's filling in the `{}` with `jobid`

sjanssen2 commented on cluster_configs/barnacle/barnacle_status.py:
```diff
@@ -0,0 +1,64 @@
+def parse_qstat(lines):
+    q_dict = {}
+
+    p = re.compile('^ +(.+?) = (.+)')
```
I found that regular expressions in python are terrible slow. Thus, if somehow possible you should avoid them if performance is important. I don't know how many lines you have to parse here, if it is < 1000 it should be fine as is

tanaes replied:
It's parsing [these status reports](https://github.com/tanaes/oecophylla/blob/1c43b6ae2084211a14a4faa4f10989d22afe0937/cluster_configs/barnacle/tests/test_data/success.txt) from qstat, so like 60-100 lines. However, it is doing it a few times per second per simultaneous running job. Do you think that's fine?

qiyunzhu commented:
Shall I merge? @tanaes

tanaes commented:
Yes please!

On Wed, Oct 25, 2017 at 2:10 PM Qiyun Zhu <notifications@github.com> wrote:

> Shall I merge? @tanaes <https://github.com/tanaes>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/oecophylla/pull/120#issuecomment-339419600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH6JADK2YkryP5YtXlUiEZ2iLJb1Twtxks5sv3j6gaJpZM4QAA_u>
> .
>
